### ISSUE TITLE
Oak-specific metadata / headers format (protobuf vs. JSON)

### ISSUE THREAD

**Issue by tiziano88** [project-oak/oak#1724]
Currently we have two versions of each Oak-specific gRPC metadata / HTTP header , one for protobuf and one for JSON encoded data.

We should discuss whether it makes sense to keep maintaining both, or should we focus on one. AFAIR the main reason for the JSON versions was to make it easier to work with JavaScript code, but I don't remember the details.

@rbehjati @juliettepretot do you remember more of the background?

**Comment by rbehjati** [comment]
I don't think the JSON version is only for JavaScript. My understanding is that we generally need the JSON format for HTTP requests, since HTTP/1.1 is currently the version we support in the HTTP server pseudo-node. REST + JSON over HTTP/1.1 is a more common pattern. 

We could discuss whether the support for protobuf encoded headers in the HTTP requests is needed. I personally, don't see the need for that.

**Comment by tiziano88** [comment]
I'm not sure there is much precedent for encoding either JSON or protobuf as HTTP headers, apart perhaps from things like JWT. In any case, I think it's mostly down to how we expect clients to populate these headers, and for that I think we would want a JS library of some sort to have an idiomatic API, and presumably that may as well generate either JSON or protobuf of the correct format.

**Comment by juliettepretot** [comment]
For reference, the original issue and discussion that led to implementing both protobuf & JSON:
https://github.com/project-oak/oak/issues/1279

It is true that encoding either as headers is not a common pattern. If we wanted things to be a bit more idiomatic we could send the labels as part of a POST payload, but I'm not sure if it matters that much. ¯\_(ツ)_/¯

JSON does make things way easier for web applications, which are otherwise forced to ship a pretty large lib to support protobuf. 

Protobuf over HTTP isn't really a common pattern, imho we can kill that option for the Oak runtime. 

I think @rbehjati might've included it because protobuf over HTTP is a pattern that my internal introspection tooling uses. But that was a pragmatic choice for that particular usecase, it certainly isn't common enough to justify its inclusion in the Oak runtime.

**Comment by tiziano88** [comment]
@juliettepretot how hard is it in practice to include the protobuf lib? I would expect we will need to come up with a JS client library for Oak at some point (performing challenge-response auth and attaching labels) so we may as well bundle compiled protobuf together with it.

**Comment by juliettepretot** [comment]
Assuming we'd stick with the official protobuf implementation the library in question would be https://www.npmjs.com/package/google-protobuf

There's two factors two consider:

### Load time
[Over 50% of users abandon a website if it takes longer than 3 seconds to load.](https://www.thinkwithgoogle.com/intl/en-154/marketing-strategies/app-and-mobile/need-mobile-speed-how-mobile-latency-impacts-publisher-revenue/)
The size of a website's JavaScript payload is an important consideration. That's because user's pay doubly for it in terms of load times. A) time spent on the actual network call, and B) time spent parsing & optimizing the downloaded JS file. 

Unfortunately `google-protobuf` is a heavy-weight, clocking in at 229.5kB after compression & minification. That's larger than the entire JS payload of many websites, and takes about 1s to download on 3G, plus time spent parsing, which depends on the speed of user's CPU. On low end Android phones that could easily be another second. In practice it means it's not really useable for users on slow phones & connections (emerging markets). That's among the reasons why the vast majoroity of Google's websites don't use protobuf.

### Bundling
Unfortunately `google-protobuf` only outputs non-standard Common-JS or `goog.require` style modules, instead of native EcmaScript modules. Defacto it means that user's need to use a build system just because of that lib. Besides some [performance implications](https://web.dev/commonjs-larger-bundles/) the ecosystem is moving to ES modules. It be good to support it for any new JS library written.

### gRPC 
If we use protobuf on the client it would be idiomatic to also use gRPC. In that case we'd need to add support to the Oak runtime to serve gRPC over HTTP/1, ala https://github.com/grpc/grpc-web


In conclusion it'd work for proof of concept type applications, but pending improvements to the protobuf JS implementation it would not be advisable for production Oak apps.

**Comment by tiziano88** [comment]
Thanks for writing this down @juliettepretot , it is useful!

It sounds like web applications are going to stick to JSON over HTTP, and native applications with protobuf over gRPC. If that's the case, then it may make sense to support labels in JSON and protobuf format, respectively, both following the protobuf schema, even if it means that web applications will just create JSON objects "by hand" rather than relying on any actual enforcement of the schema.

I think this is quite unfortunate, since it means that gRPC traffic (which is actually also HTTP) will use a different label format from plain HTTP traffic, with potential security implications if the matching logic is incorrect. Also it means that we will not be able to unify HTTP and gRPC server logic, since they will require different handling of headers, if nothing else, even though in both cases they would be HTTP headers.

In practice I think it means we can drop support for protobuf-serialized labels for the HTTP server pseudo-node, and simplify the header names and matching logic?

**Comment by juliettepretot** [comment]
@tiziano88 No problem, happy to. It is indeed unfortunate. I agree with your conclusions, web as JSON over HTTP, and native as protobuf over gRPC seems like the endgame for production apps.

That said, it's worth considering sticking with protobuf-over-grpc _for now_, if it alleviates the issues you mentioned. 

The reason is that Oak is far from being production ready in either case. And if maintaining JSON logic is costly, it's fine to use protobuf on the web for proof of concept type applications. Once people want to use Oak on the web in prod, we can invest the energy of maintaining dual protobuf + JSON logic.

**Comment by juliettepretot** [comment]
An example of this being the introspection tooling. It currently uses protobuf over http on the web because it makes the code much easier. That's not ideal, but given the state of project it's fine for now.

**Comment by rbehjati** [comment]
What is exactly the cost of maintaining the JSON logic currently?

**Comment by tiziano88** [comment]
Mental overhead and additional maintenance cost for one, but mostly I am worried about the risk of someone sending a gRPC request with a JSON label or vice versa, currently we have matching logic to ensure that only one label is present, and needs to be parsed, etc. I can see that logic getting wrong in the future as we add more functionality around it, and ending up with some confused deputy attacks or other parsing issues.

Anyways, for now it is fine, but I would like to see a solid plan (i.e. a simpler solution to maintain) longer term.

**Comment by rbehjati** [comment]
My understanding from the discussions above is that the plan for the long term would be to allow protobuf with gRPC, and JSON with HTTP. I think we should stick to that, and removed the protobuf support that we currently have for the HTTP server node. 

I am not sure I understand the risk of sending a gRPC request with a JSON label. Are you worried that a unified logic for HTTP and gRPC server nodes might lead to allowing such risky scenarios?

**Comment by tiziano88** [comment]
Exactly.

**Comment by tiziano88** [comment]
I mean, a possible solution may be to just use JSON for both gRPC and HTTP, that may actually simplify things quite a bit, even though it would not be very idiomatic for gRPC, but it would be abstracted into a client library anyways, so not a big deal.

**Comment by tiziano88** [comment]
Using JSON for both would actually be similar to other approaches like JWT (which has other issues, but let's leave that aside for now), so it's not really unheard of.

**Comment by tiziano88** [comment]
BTW similar arguments apply to the challenge / response authentication mechanism, of which it would be good to have only one version too.
